Update education rule description documentation with more details (#1610)
SonarSource/rspec: docs/description.adoc

include::section_names.adoc[]

Most third-level and fourth-level titles (`+=== Title t3+` and `+==== Title t4+`) are not checked, the only type of rule description for which some of these

> titles are checked are the progressive education rules (see below).

Reviewer: ```suggestion
titles are checked are the education format rules (see below).
```

> 

Reviewer: the line above `The following are the only allowed second level titles for H2:` is only valid for the standard and hotspot formats. If we detect a header from the new format, then we allow only another set of headers: https://github.com/SonarSource/rspec/blob/fee0463e41f664da9121f5ca6c644a5a21c2b618/rspec-tools/rspec_tools/validation/description.py#L33-L40

. See
. See Also

=== 3. Education Format

Finally, the last format is the one related to the Progressive Education.
This new format has a completely different structure, with additional subsections and information,
to be as comprehensible as possible for the users.

This new structure is defined as follows:

* Short description (no title)

> // This needs to be kept in sync with the [maps in the validation script](https://github.com/SonarSource/rspec/blob/3fb33d25c7bcfcde5a2426cb7177389cdcb6f5eb/rspec-tools/rspec_tools/validation/description.py#L32-L42).

Reviewer: ```suggestion
// This needs to be kept in sync with the [maps in the validation script](https://github.com/SonarSource/rspec/blob/master/rspec-tools/rspec_tools/validation/description.py#L33-L40).
```

I've updated the link to point at a commit from master instead of the feature branch

. See
. See Also

=== 3. Education Format

Finally, the last format is the one related to the Progressive Education.

> This new format has a completely different structure, with additional subtitles and information,

Reviewer: ```suggestion
This new format has a completely different structure, with additional subsections and information,
```

.. Conference Presentations
.. Standards

Where section 3 can be repeated multiple times when the rule description is defined per-framework.

The sections and subsections for this format are defined as follows:

* Short description (no title) [Optional]

>     Short introduction/summary of the topic, only a few sentences.

Reviewer: Maybe not use a code block for this as you need to scroll in the rendered version which is not convenient to read docs?

Reviewer: see example: https://github.com/SonarSource/rspec/blob/czurn/update-documentation-for-education/docs/description.adoc#3-education-format

Goal: The title (or message) of a rule might not always be clear due to its shortness, and this should make it clear to our users what the issue is about. Experienced developers will immediately understand what it is about, and developers not yet familiar with the issue at hand can dig deeper into the “Why is this an issue?” section.
+
* *Why is this an issue?* (level 2 title) [Mandatory]
+
Start at the basics and go into more depth to explain the concepts behind this type of issue. This is most likely the place where a lot of the content will be.
Goal: Understand the concepts behind an issue.
+
* *What is the potential impact?* (level 3 title) [Optional]
+
What is the risk for me? What can go wrong?
This section can include multiple concrete threats as well.

> Goal: Our users should understand the impact of this issue on their project.

Reviewer 2: In this location, the AppSec bubble generally worked like the [following example for SQL injections](https://github.com/SonarSource/rspec/blob/a51217c6d91abfa5e1d77d0ae0843e3903adf2d0/rules/S3649/impact.adoc):

---

=== What is the potential impact?

***[Context:  - Initial context  - Situation on how an attacker exploits the vulnerability ]***
In the context of a web application that is vulnerable to SQL injection: +
After discovering the injection, attackers inject data into the vulnerable
field to execute malicious commands in the affected databases.

***[Transition sentence that we added everywhere]***
Below are some real-world scenarios that illustrate some impacts of an attacker
exploiting the vulnerability.

***[Confidentiality/Privilege escalation/Integrity Impact - most widespread first]***
==== Identity spoofing and data manipulation

A malicious database query enables privilege escalation or direct data leakage
from one or more databases. This threat is the most widespread impact.

***[Availability Impact]***
==== Data deletion and denial of service

The malicious query makes it possible for the attacker to delete data in the
affected databases. +
This threat is particularly insidious if the attacked organization does not
maintain a disaster recovery plan (DRP).

***[Impact a little harder to grasp for everyone but still totally relevant]***
==== Chaining DB injections with other vulnerabilities

Attackers who exploit SQL injections rely on other vulnerabilities to maximize
their profits. +
Most of the time, organizations overlook some defense in depth measures because
they assume attackers cannot reach certain points in the infrastructure. This
misbehavior can lead to multiple attacks with great impact:

* When secrets are stored unencrypted in databases: Secrets can be exfiltrated and lead to compromise of other components.
* If server-side OS and/or database permissions are misconfigured, injection can lead to remote code execution (RCE).
** See https://blog.sonarsource.com/exploiting-hibernate-injections/[our article on this topic]

Author: I've added a link to the example in the documentation.
I'm not sure how to add this approach in the documentation though... I m open to ideas :D Coz I think it's interesting

Reviewer 2: WDYT of a subsection here on how to do it that would be a list that looks like this (this is a draft with ideas):
1. State how a realistic situation might appear (a user doing this, a developer doing that) - it might repeat what was already said in the previous paragraph, but with another way of explaining (explaining it twice differently might improve understanding)
2. [Optional ?] Add a transition sentence, such as "Here are some real-world scenarios that illustrate some impacts of [an attacker exploiting this vulnerability. | a colleague falling in the trap of this code smell. | a triggering of this bug.]

## Vulns
For Security issues: Discuss impacts to confidentiality, integrity, and availability (when relevant) as a first basis, other sections can include privilege escalation or vulnerability chaining. Try to be as realistic as possible.

## Code smells
What does this code issue of readability/maintainability/understandability can do to a code base ?

Here is a real-world example, `java:S1188` ("Anonymous classes should not have too many lines"), written how I would assume it would be written (or kinda): 

---
Anonymous classes make the code clearer and have the advantage that they can be declared and instantiated at the same time. They are like local classes, except that they have no name.
Anonymous classes are often used in graphical user interface (GUI) applications and are ideal for implementing an interface that contains only a few methods.

### What is the potential impact?
The context: a few years after coding several functions that interact with the class, the users who programmed them have left the company without having time to write the documentation. A new colleague needs to change the behavior of the class or convert it to a named class because the scope of this class went beyond what was originally intended, or because this class exists multiple times in the code base.

The cognitive complexity of reading an overly long anonymous class makes this task either lengthy, tedious, or error-prone, depending on the code base.

### Functional Bugs
Rewriting/using this class can lead to unintended bugs due to the original environment in which it was declared, such as hidden dependencies on local data.

### Loss of functions
Rewriting or using this class without understanding its intended behavior or known issues may result in new truncated branches and remove functions because they can no longer be maintained.

### Burnout
The newcomer rewrites or reuses this class without understanding its intended behavior. They cry, think they are not worth the job, and burn out because of you within two days of joining the company. Congratulations

---

## Bugs
What would keeping this bug in the code would provoke when triggered / In the long run? Here I would mostly see variants of availability impacts. Tell me if you want me to illustrate this with a real bug example with my point of view

+
* *How to fix it* or  *How to fix it in `{Framework Display Name}`* (level 2 title) [the title depends on whether the description is generic or framework-specifc. See examples below.]
+
This section consists of one or multiple fixes for this type of issue (`Noncompliant code example` vs. `Compliant solution`). There can be multiple fixes for different libraries and/or frameworks.
Goal: Get an idea of how this issue can be fixed for my project/framework, why this works, what to look out for, and also how to continue improving on this topic.
+
* *How does this work?* (level 3 title) [Optional]
+
Explain why this fixes the problem.
+
* *Pitfalls* (level 3 title) [Optional]
+

> One or multiple pitfalls to take into account when working on fixing this issue.

Reviewer 2: ```suggestion
One or multiple pitfalls to take into account when working on fixing this issue. [_Example_](https://github.com/SonarSource/rspec/blob/a51217c6d91abfa5e1d77d0ae0843e3903adf2d0/rules/S6096/common/pitfalls/partial-path-traversal.adoc)
```

+
* *Going the extra mile* (level 3 title) [Optional]
+
Even though the issue might be fixed, most of the time there can be way/s to further improve on this issue or to harden your project.

> The subsection should be concise.

Reviewer 2: ```suggestion
The subsection should be concise. [_Example_](https://github.com/SonarSource/rspec/blob/a51217c6d91abfa5e1d77d0ae0843e3903adf2d0/rules/S5131/common/extra-mile/csp.adoc)
```

+
* *Resources* (level 2 title) [Optional]
+

> Include resources if our users want to dig even deeper, that can be presented in the different categories.

Reviewer 2: This location also includes the "Clean Code" concepts/principles, which we previously called "Secure Coding principles". Is it worth mentioning here, as FYI?

Author: I'm not entirely sure. Is this something we are expecting the others to use?
TBH even for me this concept of "Clean Code principles" is not very clear. How can we add one? Update one?
I think it'd be best, for now, to keep it out, to avoid confusion, and add it later. WDYT?

Reviewer 2: Yes, agreed! It starts to get out of the scope of rspec if we discuss this here.

> Include resources if our users want to dig even deeper, that can be presented in the different categories.

Reviewer 2: ```suggestion
Include resources if our users want to dig even deeper, that can be presented in the different categories. [_Example_](https://github.com/SonarSource/rspec/tree/a51217c6d91abfa5e1d77d0ae0843e3903adf2d0/rules/S5131/common/resources)
```

Goal: Allow the user to dig deeper by providing a curated list of resources.
+
* *Documentation* (level 3 title) [Optional]
* *Articles & Blog Posts* (level 3 title) [Optional]
* *Conference Presentations* (level 3 title) [Optional]
* *Standards* (level 3 title) [Optional]

Note that most sections and subsections are optional, only the `Why is this an issue?` and `How to fix it`/ `How to fix it in {Framework Display Name}` main
sections are mandatory.

Content of the section "_3. How to fix it_ / _How to fix it in {Framework Display Name}_" can either be generic or framework specific.

When the content is generic, the "_How to fix it_" title must be used, and the section should only appear once. Example:
....
== Why is this an issue?
Explanation of why this is bad.

== How to fix it

The fix is actually to simply do something.

==== Noncompliant code example
[source,diff-id=1,diff-type=noncompliant]
----
var myExample;
----

==== Compliant solution
[source,diff-id=1,diff-type=compliant]
----
var myExample = 0;
----

=== How does this work?
We added something.

== Resources
=== Documentation
http-address-of-documentation[My doc name]

....

> When the content is framework-specific, one or more "_3. How to fix it in `{Framework Display Name}`_" sections (with their respective subsections) can be present.

Reviewer 3: ```suggestion
When the content is framework-specific, one or more "_3. How to fix it in `{Framework Display Name}`_" sections (with their respective subsections) must be present.
```
Given that the rule must have at least one "_How to fix it ..._" section, and the generic one is only allowed in rules that are framework-agnostic

> When the content is framework-specific, one or more "_3 How to fix it in `{Framework Display Name}`_" sections (with their respective subsections) can be present.

Reviewer: ```suggestion
When the content is framework-specific, one or more "_3. How to fix it in `{Framework Display Name}`_" sections (with their respective subsections) can be present.
```

** Conference presentations
** Standards

Where the `How to fix it in {Framework Display Name}` section can be repeated multiple times when the rule description is defined per-framework.

The sections and subsections for this format are defined as follows:

* Short description (no title) [Optional]
+

> Short introduction/summary of the topic, only a few sentences.

Reviewer: ```suggestion
A short introduction/summary of the topic, only a few sentences.
```

> Goal: The title (or message) of a rule might not always be clear due to its shortness, and this should make it clear to our users what the issue is about. Experienced developers will immediately understand what it is about, and developers not yet familiar with the issue at hand can dig deeper into the “Why is this an issue?” section.

Reviewer: ```suggestion
Goal: The title (or message) of a rule might not always be clear due to its shortness, and this should make it clear to our users what the issue is about. Experienced developers will immediately understand what it is about, and developers not yet familiar with the issue at hand can dig deeper into the `Why is this an issue?` section.
```

... Generic and JSP-specific pitfalls to avoid when fixing the issue...
....

Ideally, by convention and for maintainability, each framework _How to fix it_ section will be defined in separate files.
Ex:
....
 == Why is this an issue?
 ... Explanation ...

 # How to fix it sections

 include::./how-to-fix-it/framework-1.adoc[]

 include::./how-to-fix-it/framework-2.adoc[]

 == Resources
 === Documentation
 http-address-of-documentation[My doc name]
....

> Note that each framework-specific _How to fix it_ subsection must start with an H2 title following the given format:
> `How to fix it in (?:(?:an|a|the)\\s)?(.*)`.

Reviewer 3: ```suggestion
Note that each framework-specific _How to fix it_ subsection must start with an H2 title (`== `) following the given format:
`How to fix it in (?:(?:an|a|the)\\s)?(.*)`.
```
Is it useful to cite a literal regular expression here? IMO, you can put a reference to the actual regular expression used in the check, and here write a human-friendly version, e.g. "`How to fix it in [an|a|the] {Framework name}` or `How to fix it`"

Author: Good point. Thanks :+1:

Note that each framework-specific _How to fix it?_ subsection must start with an H2 subtitle following the given format:
Note that each framework-specific _How to fix it_ subsection must start with an H2 title following the given format:
`How to fix it in (?:(?:an|a|the)\\s)?(.*)`.
This is important, as this format will be expected by the analyzers when loading the rule content to recognize the different subsections.
Furthermore, the display name of the framework has to match an allowed framework
This is important, as this format is used by the CI to check the validity of the title, and the same format will be
expected by the analyzers when loading the rule content to recognize the different subsections.
Furthermore, to avoid having the same framework used with different names, the display name of the framework has to match an allowed framework
display name, as defined in <<allowed_framework_names.adoc#,this allowed framework names file>>.

Finally, additional information regarding each of the sections and subsections present in the educational rule description
can be found https://docs.google.com/document/d/1caqBXTO1r7vftWrNXt_aNVe4rHcB5C6u6oKMBFjbzwo[in this document].

> 

Reviewer 2: The code samples mentionned after do not mention the new attributes of the diff view:

```
[php,diff-id=1,diff-type=noncompliant]
```

Author: Good point. I added a note and a section below explaining these 2 new attributes for code examples :+1:

Ideally, by convention and for maintainability, each framework _How to fix it_ section will be defined in separate files.
Ex:
....
 == Why is this an issue?
 ... Explanation ...

 # How to fix it sections

 include::./how-to-fix-it/framework-1.adoc[]

 include::./how-to-fix-it/framework-2.adoc[]

 == Resources
 === Documentation
 http-address-of-documentation[My doc name]
....

Note that each framework-specific _How to fix it_ subsection must start with an H2 title following the given format:

> `== How to fix it in [an|a|the] {Framework name}`.

Reviewer: ```suggestion
`== How to fix it in [an|a|the]? {Framework name}`.
```

the article is optional, no?

In case no language is appropriate for a code block (for example shared examples between multiple languages), you can use `text` as the language.

=== Diff view

Additionally, you can also use two attributes to let the products know your code examples should be highlighted with a diff view when possible
(showing the changes in the code examples as red/green).
These attributes are optional and if a product does not yet support the diff view feature, these attributes will simply be ignored.

These attributes are `diff-id=X` and `diff-type=[noncompliant|compliant]`. The `diff-id` attributes describe which code examples should
be compared together, and the `diff-type` attribute explain how it should be displayed `Noncompliant` (red) vs. `Compliant` (green).

> A single and unique diff-id should be used only once for each type of code example in the description of a rule.

Reviewer: ```suggestion
A single and unique diff-id should be used only once for each type of code example as shown in the description of a rule.
```
